### DISCUSSION TITLE
Native UUID type support

### DISCUSSION THREAD

**Discussion by sas1024** [go-pg/pg#1866] [Q&A] [answered]
Is it possible to add native postgresql UUID type support for creating table over model with CreateTable func?
https://www.postgresql.org/docs/9.6/static/datatype-uuid.html

**Comment by vmihailenco** [comment]
Did you try https://github.com/google/uuid or https://github.com/satori/go.uuid?

**Comment by sas1024** [comment]
I tried https://github.com/satori/go.uuid - it supports UUIDv5.

**Comment by vmihailenco** [comment]
For `CreateTable` you can add field annotation

```
type Article struct {
    Id uuid.UUID `sql:",type:uuid"`
}
```

Later go-pg can be improved to treat `[16]byte` as uuid (but I am not sure if it should be)...

**Comment by sas1024** [comment]
Thank you very much!

**Comment by satish71** [comment, marked as answer]
I am new to go and go pg. Here is what I had learnt the hard way

1. UUID is not build in support, you need add an _UUID extension_ to a postgres database. this is a contrib. extension that's reason you need to explicitly create it.

`CREATE EXTENSION IF NOT EXISTS "uuid-ossp"; `

2. Extension are per database. if you have create a new database. You need to repeat the above procedure   
3. Use the latest version of go-pg as earlier version had problem
4. Ensure above imports, please pay attention the package references 'pg' and 'uuid'. We would using them.

```
import (
	pg "github.com/go-pg/pg/v10"
	"github.com/go-pg/pg/v10/orm"
	uuid "github.com/satori/go.uuid"
)
```
5. You need now tell go-pg to correctly map your struct to postgres sql type.  if you don't pay attention to this details especially usage of 'pg' and 'type:uuid. most of the information available on net refers 'sql' and using will result into creation of table with type 'bytea' instead of 'uuid'

 you need to use **'pg'** as a package reference not _'sql'_
```
type Story struct {
	Id       uuid.UUID `pg:",pk,type:uuid default uuid_generate_v4()"`
	Title    string
	AuthorId uuid.UUID `pg:",type:uuid DEFAULT uuid_generate_v4()"`
}
```

**Reply by vmihailenco** [reply]
:+1:  I've added https://pg.uptrace.dev/uuid/ based on your instructions.
